In [1]:
import pyodbc     
from dotenv import dotenv_values  
import pandas as pd
import warnings 

warnings.filterwarnings('ignore')

# Load environment variables from .env file into a dictionary
environment_variables = dotenv_values('.env')

server = 'dap-projects-database.database.windows.net'
database = 'dapDB'
username = 'LP2_project'
password = 'Stat$AndD@t@Rul3'
table_name = 'dbo.LP2_Telco_churn_first_3000'

# Connection string
conn_str = f'DRIVER={{SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password}'

# Connect to the database
conn = pyodbc.connect(conn_str)

# Query the database and load data into a DataFrame
query = "SELECT * FROM LP2_Telco_churn_first_3000"
data1 = pd.read_sql(query, conn)

data1.head(5)


customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG  Female          False     True       False       1   
1  5575-GNVDE    Male          False    False       False      34   
2  3668-QPYBK    Male          False    False       False       2   
3  7795-CFOCW    Male          False    False       False      45   
4  9237-HQITU  Female          False    False       False       2   

   PhoneService MultipleLines InternetService OnlineSecurity  ...  \
0         False          None             DSL          False  ...   
1          True         False             DSL           True  ...   
2          True         False             DSL           True  ...   
3         False          None             DSL           True  ...   
4          True         False     Fiber optic          False  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0            False       False       False           False  Month-to-month   
1             True       False       False           False        One year   
2            False       False       False           False  Month-to-month   
3             True        True       False           False        One year   
4            False       False       False           False  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0             True           Electronic check      29.850000     29.850000   
1            False               Mailed check      56.950001   1889.500000   
2             True               Mailed check      53.849998    108.150002   
3            False  Bank transfer (automatic)      42.299999   1840.750000   
4             True           Electronic check      70.699997    151.649994   

   Churn  
0  False  
1  False  
2   True  
3  False  
4   True  

[5 rows x 21 columns]

In [2]:
# #Saving the dataset
# data1.to_csv('Dataset 1', index=False)

In [2]:
# saving data1 as csv file
data1 = pd.DataFrame(data1)

# Save DataFrame to CSV
data1.to_csv('Datasets 1.csv', index=False)

In [3]:

# Raw URL of the dataset file
url = 'https://raw.githubusercontent.com/Azubi-Africa/Career_Accelerator_LP2-Classifcation/main/LP2_Telco-churn-second-2000.csv'

# Read the dataset into a DataFrame
data2= pd.read_csv(url)

data2.head(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  5600-PDUJF    Male              0      No         No       6          Yes   
1  8292-TYSPY    Male              0      No         No      19          Yes   
2  0567-XRHCU  Female              0     Yes        Yes      69           No   
3  1867-BDVFH    Male              0     Yes        Yes      11          Yes   
4  2067-QYTCF  Female              0     Yes         No      64          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0                No             DSL             No  ...               No   
1                No             DSL             No  ...              Yes   
2  No phone service             DSL            Yes  ...              Yes   
3               Yes     Fiber optic             No  ...               No   
4               Yes     Fiber optic             No  ...              Yes   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes          No              No  Month-to-month              Yes   
1         Yes          No              No  Month-to-month              Yes   
2          No          No             Yes        Two year              Yes   
3          No          No              No  Month-to-month              Yes   
4         Yes         Yes             Yes  Month-to-month              Yes   

             PaymentMethod MonthlyCharges  TotalCharges Churn  
0  Credit card (automatic)          49.50         312.7    No  
1  Credit card (automatic)          55.00        1046.5   Yes  
2  Credit card (automatic)          43.95        2960.1    No  
3         Electronic check          74.35         834.2   Yes  
4         Electronic check         111.15        6953.4    No  

[5 rows x 21 columns]

In [4]:
# #Saving the dataset
# data2.to_csv('Dataset 2', index=False)

In [4]:
# saving data2 as csv file
data2 = pd.DataFrame(data2)

# Save DataFrame to CSV
data2.to_csv('Datasets 2.csv', index=False)

In [5]:
data2.head(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  5600-PDUJF    Male              0      No         No       6          Yes   
1  8292-TYSPY    Male              0      No         No      19          Yes   
2  0567-XRHCU  Female              0     Yes        Yes      69           No   
3  1867-BDVFH    Male              0     Yes        Yes      11          Yes   
4  2067-QYTCF  Female              0     Yes         No      64          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0                No             DSL             No  ...               No   
1                No             DSL             No  ...              Yes   
2  No phone service             DSL            Yes  ...              Yes   
3               Yes     Fiber optic             No  ...               No   
4               Yes     Fiber optic             No  ...              Yes   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes          No              No  Month-to-month              Yes   
1         Yes          No              No  Month-to-month              Yes   
2          No          No             Yes        Two year              Yes   
3          No          No              No  Month-to-month              Yes   
4         Yes         Yes             Yes  Month-to-month              Yes   

             PaymentMethod MonthlyCharges  TotalCharges Churn  
0  Credit card (automatic)          49.50         312.7    No  
1  Credit card (automatic)          55.00        1046.5   Yes  
2  Credit card (automatic)          43.95        2960.1    No  
3         Electronic check          74.35         834.2   Yes  
4         Electronic check         111.15        6953.4    No  

[5 rows x 21 columns]

In [6]:
data1.head(5)

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG  Female          False     True       False       1   
1  5575-GNVDE    Male          False    False       False      34   
2  3668-QPYBK    Male          False    False       False       2   
3  7795-CFOCW    Male          False    False       False      45   
4  9237-HQITU  Female          False    False       False       2   

   PhoneService MultipleLines InternetService OnlineSecurity  ...  \
0         False          None             DSL          False  ...   
1          True         False             DSL           True  ...   
2          True         False             DSL           True  ...   
3         False          None             DSL           True  ...   
4          True         False     Fiber optic          False  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0            False       False       False           False  Month-to-month   
1             True       False       False           False        One year   
2            False       False       False           False  Month-to-month   
3             True        True       False           False        One year   
4            False       False       False           False  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0             True           Electronic check      29.850000     29.850000   
1            False               Mailed check      56.950001   1889.500000   
2             True               Mailed check      53.849998    108.150002   
3            False  Bank transfer (automatic)      42.299999   1840.750000   
4             True           Electronic check      70.699997    151.649994   

   Churn  
0  False  
1  False  
2   True  
3  False  
4   True  

[5 rows x 21 columns]

In [7]:
data1.describe()

tenure  MonthlyCharges  TotalCharges
count  3000.000000     3000.000000   2995.000000
mean     32.527333       65.347400   2301.278315
std      24.637768       30.137053   2274.987884
min       0.000000       18.400000     18.799999
25%       9.000000       35.787499    415.250000
50%      29.000000       70.900002   1404.650024
75%      56.000000       90.262501   3868.725098
max      72.000000      118.650002   8564.750000

In [8]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        3000 non-null   object 
 1   gender            3000 non-null   object 
 2   SeniorCitizen     3000 non-null   bool   
 3   Partner           3000 non-null   bool   
 4   Dependents        3000 non-null   bool   
 5   tenure            3000 non-null   int64  
 6   PhoneService      3000 non-null   bool   
 7   MultipleLines     2731 non-null   object 
 8   InternetService   3000 non-null   object 
 9   OnlineSecurity    2349 non-null   object 
 10  OnlineBackup      2349 non-null   object 
 11  DeviceProtection  2349 non-null   object 
 12  TechSupport       2349 non-null   object 
 13  StreamingTV       2349 non-null   object 
 14  StreamingMovies   2349 non-null   object 
 15  Contract          3000 non-null   object 
 16  PaperlessBilling  3000 non-null   bool   


In [9]:
# showing all the null values in the columns of the data from the server. 9 columns have null values in them.

data1.isnull().sum()

customerID            0
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines       269
InternetService       0
OnlineSecurity      651
OnlineBackup        651
DeviceProtection    651
TechSupport         651
StreamingTV         651
StreamingMovies     651
Contract              0
PaperlessBilling      0
PaymentMethod         0
MonthlyCharges        0
TotalCharges          5
Churn                 1
dtype: int64

In [10]:
data2.describe()

SeniorCitizen       tenure  MonthlyCharges
count    2043.000000  2043.000000     2043.000000
mean        0.168380    32.649046       64.712555
std         0.374295    24.376248       29.970010
min         0.000000     0.000000       18.550000
25%         0.000000     9.000000       35.825000
50%         0.000000    30.000000       70.250000
75%         0.000000    55.000000       89.625000
max         1.000000    72.000000      118.350000

In [11]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2043 entries, 0 to 2042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        2043 non-null   object 
 1   gender            2043 non-null   object 
 2   SeniorCitizen     2043 non-null   int64  
 3   Partner           2043 non-null   object 
 4   Dependents        2043 non-null   object 
 5   tenure            2043 non-null   int64  
 6   PhoneService      2043 non-null   object 
 7   MultipleLines     2043 non-null   object 
 8   InternetService   2043 non-null   object 
 9   OnlineSecurity    2043 non-null   object 
 10  OnlineBackup      2043 non-null   object 
 11  DeviceProtection  2043 non-null   object 
 12  TechSupport       2043 non-null   object 
 13  StreamingTV       2043 non-null   object 
 14  StreamingMovies   2043 non-null   object 
 15  Contract          2043 non-null   object 
 16  PaperlessBilling  2043 non-null   object 


In [13]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        3000 non-null   object 
 1   gender            3000 non-null   object 
 2   SeniorCitizen     3000 non-null   bool   
 3   Partner           3000 non-null   bool   
 4   Dependents        3000 non-null   bool   
 5   tenure            3000 non-null   int64  
 6   PhoneService      3000 non-null   bool   
 7   MultipleLines     2731 non-null   object 
 8   InternetService   3000 non-null   object 
 9   OnlineSecurity    2349 non-null   object 
 10  OnlineBackup      2349 non-null   object 
 11  DeviceProtection  2349 non-null   object 
 12  TechSupport       2349 non-null   object 
 13  StreamingTV       2349 non-null   object 
 14  StreamingMovies   2349 non-null   object 
 15  Contract          3000 non-null   object 
 16  PaperlessBilling  3000 non-null   bool   


In [12]:
# showing all the null values in the columns of the data from the github
# No null values found in data2

data2.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

# Clean Columns
In both Datasets SeniorCitizen, Partner, Dependents, PhoneService, PaperlessBilling, TotalCharges all have different datatypes. This has to be changed to the same datatypes. 

MultipleLines in data1 has 269 missing values and needs to be cleaned.

OnlineSecurity, OnlineBackup, DeviceProtection, StreamingTV, TechSupport have 651 missing values in data1

TotalCharges has 5 missing values in data1 and datatypes are different.
Churn has 1 missing value in data1

Contract both
PaymentMethod both

# SeniorCitizen
There are no missing data present but the datatype is different. data1 is a boolean while data2 is made up of integers

In [14]:
# check unique values in data1

data1['SeniorCitizen'].unique()

array([False,  True])

In [15]:
# check unique values in data2

data2['SeniorCitizen'].unique()

array([0, 1], dtype=int64)

In [17]:
# Convert the binary column to boolean datatype

data2['SeniorCitizen'] = data2['SeniorCitizen'].astype(bool)

In [18]:
# check unique values in data2

data2['SeniorCitizen'].unique()

array([False,  True])

# Partner
Convert data2['Partner'] to boolean datatype to reflect Data understanding documentation

Whether the customer has a partner or not (Yes, No)

In [19]:
# preview first 5 data in 'partner' column in data2

data2['Partner'].head()

0     No
1     No
2    Yes
3    Yes
4    Yes
Name: Partner, dtype: object

In [20]:
#  check unique data in partner column in data2

data2['Partner'].unique()

array(['No', 'Yes'], dtype=object)

In [21]:
#Replace 'Yes' or 'No' with True or False

data2['Partner'].replace('Yes', True, inplace=True)
data2['Partner'].replace('No', False, inplace=True)

data2['Partner'].head()

0    False
1    False
2     True
3     True
4     True
Name: Partner, dtype: bool

# Dependents
convert data2['Dependents'] to boolean

In [23]:
# # preview first 5 data in 'Dependents' column in data2
data2['Dependents'].head()

0     No
1     No
2    Yes
3    Yes
4     No
Name: Dependents, dtype: object

In [25]:
# Replace 'Yes' or 'No' with True or False to change datatype to boolean

data2['Dependents'].replace('Yes', True, inplace=True)
data2['Dependents'].replace('No', False, inplace=True)

data2['Dependents'].head()

0    False
1    False
2     True
3     True
4    False
Name: Dependents, dtype: bool

# PhoneService
convert data2['PhoneService'] to boolean

In [26]:
# preview first 5 data in PhoneService column

data2['PhoneService'].head()

0    Yes
1    Yes
2     No
3    Yes
4    Yes
Name: PhoneService, dtype: object

In [27]:
#Replace 'Yes' or 'No' with True or False

data2['PhoneService'].replace('Yes', True, inplace=True)
data2['PhoneService'].replace('No', False, inplace=True)

data2['PhoneService'].head()

0     True
1     True
2    False
3     True
4     True
Name: PhoneService, dtype: bool

# MultipleLines
data1 has 269 missing values, impute with mode

In [28]:
#confirm number of missing values

data1['MultipleLines'].isnull().sum()

269

In [29]:
#Check unique values in MultipleLines

data1['MultipleLines'].unique()

array([None, False, True], dtype=object)

In [30]:
#Count unique values

data1['MultipleLines'].value_counts()

MultipleLines
False    1437
True     1294
Name: count, dtype: int64

In [31]:
#impute missing values with mode
mode_multiplelines = data1['MultipleLines'].mode()[0]

data1['MultipleLines'].fillna(mode_multiplelines, inplace=True)
data1['MultipleLines'].isnull().sum()

0

In [32]:
data1['MultipleLines'].value_counts()

MultipleLines
False    1706
True     1294
Name: count, dtype: int64

In [40]:
# confirm unique values in data1
data1['MultipleLines'].unique()

array([False,  True])

In [34]:
data2['MultipleLines'].head(8)

0                  No
1                  No
2    No phone service
3                 Yes
4                 Yes
5                  No
6                 Yes
7                  No
Name: MultipleLines, dtype: object

In [35]:
# checking the count of the unique values MultipleLines in data2

data2['MultipleLines'].value_counts()

MultipleLines
No                  987
Yes                 836
No phone service    220
Name: count, dtype: int64

In [36]:
# we assume that an answer of 'No phone service' is the same as No, since a customer without phone service will have no need for multiplelines

# Replace 'No phone service' with 'No' in the 'MultipleLines' column
data2['MultipleLines'].replace('No phone service', 'No', inplace=True)
data2['MultipleLines'].value_counts()

MultipleLines
No     1207
Yes     836
Name: count, dtype: int64

In [41]:
#confirm number of missing values

data2['MultipleLines'].isnull().sum()

0

In [42]:
# convert datatype to boolean in data2

data2['MultipleLines'].replace('Yes', True, inplace=True)
data2['MultipleLines'].replace('No', False, inplace=True)
data2['MultipleLines'].head()

0    False
1    False
2    False
3     True
4     True
Name: MultipleLines, dtype: bool

In [43]:
# confirm unique values forMultipleLines in data2

data2['MultipleLines'].unique()

array([False,  True])

# OnlineSecurity
 OnlineSecurity -- Whether the customer has online security or not (Yes, No, No Internet)

In [44]:
# checking for unique values in data1 
data1['OnlineSecurity'].unique()

array([False, True, None], dtype=object)

In [45]:
# value count of OnlineSecurity
data1['OnlineSecurity'].value_counts()

OnlineSecurity
False    1478
True      871
Name: count, dtype: int64

In [46]:
# checking for missing values in  data1 
data1['OnlineSecurity'].isnull().sum()

651

In [47]:
# we assume the missing data is a response to the third option 'No Internet'. Hence will fill missing with 'No Internet'
# Impute missing values with 'No Internet'.


data1['OnlineSecurity'].fillna('No Internet', inplace=True)
data1['OnlineSecurity'].value_counts()

OnlineSecurity
False          1478
True            871
No Internet     651
Name: count, dtype: int64

In [48]:
# checking for data2 unique values
data2['OnlineSecurity'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [49]:
data2['OnlineSecurity'].value_counts()

OnlineSecurity
No                     1029
Yes                     585
No internet service     429
Name: count, dtype: int64

In [50]:
# Replace 'No internet service' with 'No Internet' in the 'OnlineSecurity' column in data2

data2['OnlineSecurity'].replace('No internet service', 'No Internet', inplace=True)
data2['OnlineSecurity'].value_counts()

OnlineSecurity
No             1029
Yes             585
No Internet     429
Name: count, dtype: int64

# OnlineBackup
OnlineBackup -- Whether the customer has online backup or not (Yes, No, No Internet)

OnlineBackup has 651 missing values in data1


In [55]:
data1['OnlineBackup'].isnull().sum()

651

In [57]:
# checking unique values for data1['OnlineBackup']
data1['OnlineBackup'].unique()

array([True, False, None], dtype=object)

In [70]:
# Replace 'False' with 'No' and 'True' with 'Yes' the 'OnlineBackup' column
# Fill missing values with 'No Internet'

data1['OnlineBackup'].replace(True, 'Yes', inplace=True)
data1['OnlineBackup'].replace(False, 'No', inplace=True)

# data1['OnlineBackup'].fillna('No Internet', inplace=True)
data1['OnlineBackup'].value_counts()

OnlineBackup
No             1320
Yes            1029
No Internet     651
Name: count, dtype: int64

In [71]:
# checking unique values for data2['OnlineBackup']
data2['OnlineBackup'].unique()

array(['No', 'Yes', 'No Internet'], dtype=object)

In [72]:
# Replace 'No internet service' with 'No Internet'

# data2['OnlineBackup'].replace('No internet service', 'No Internet', inplace=True)
data2['OnlineBackup'].unique()

array(['No', 'Yes', 'No Internet'], dtype=object)

# DeviceProtection
DeviceProtection has 651 missing values in data1
DeviceProtection -- Whether the customer has device protection or not (Yes, No, No internet service)

In [61]:
data1['DeviceProtection'].unique()

array([False, True, None], dtype=object)

In [62]:
data1['DeviceProtection'].isnull().sum()

651

In [63]:
# Replace 'False' with 'No' and 'True' with 'Yes' in the 'DeviceProtection' column for data1 
# Fill missing values with 'No Internet'

data1['DeviceProtection'].replace(True, 'Yes', inplace=True)
data1['DeviceProtection'].replace(False, 'No', inplace=True)

data1['DeviceProtection'].fillna('No internet service', inplace=True)
data1['DeviceProtection'].value_counts()

DeviceProtection
No                     1296
Yes                    1053
No internet service     651
Name: count, dtype: int64

In [64]:
data2['DeviceProtection'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

# TechSupport 
TechSupport has 651 missing values in data1
Whether the customer has tech support or not (Yes, No, No internet)

In [73]:
# checking for unique values in data1['TechSupport']
data1['TechSupport'].unique()

array([False, True, None], dtype=object)

In [74]:
data1['TechSupport'].value_counts()

TechSupport
False    1476
True      873
Name: count, dtype: int64

In [76]:
# Replace 'False' with 'No' and 'True' with 'Yes' in the 'TechSupport' column for data1 
# Fill missing values with 'No Internet'

data1['TechSupport'].replace(True, 'Yes', inplace=True)
data1['TechSupport'].replace(False, 'No', inplace=True)

data1['TechSupport'].fillna('No internet', inplace=True)
data1['TechSupport'].value_counts()

TechSupport
No             1476
Yes             873
No internet     651
Name: count, dtype: int64

In [77]:
# confirming null vlaues
data1['TechSupport'].isnull().sum()

0

In [78]:
# checking unique values in data2['TechSupport']
data2['TechSupport'].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [79]:
# Replace 'No internet service' with 'No internet' in the 'TechSupport' column for data2

data2['TechSupport'].replace('No internet service', 'No internet', inplace=True)
data2['TechSupport'].unique()

array(['Yes', 'No', 'No internet'], dtype=object)

# StreamingTV
StreamingTV
Whether the customer has streaming movies or not (Yes, No, No Internet service)

In [80]:
data1['StreamingMovies'].unique()

array([False, True, None], dtype=object)

In [81]:
data1['StreamingMovies'].replace(True, 'Yes', inplace=True)
data1['StreamingMovies'].replace(False, 'No', inplace=True)

data1['StreamingMovies'].fillna('No internet service', inplace=True)
data1['StreamingMovies'].value_counts()

StreamingMovies
Yes                    1199
No                     1150
No internet service     651
Name: count, dtype: int64

In [83]:
data1['StreamingMovies'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [82]:
data2['StreamingMovies'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

# PaperlessBilling
In data1 PaperlessBilling has boolean as its datatype and object as its datatype in data2 

Whether the customer has paperless billing or not (Yes, No). So datatypes should aall be in objects (Yes, No)
No missing values detected 

In [84]:
data1['PaperlessBilling'].unique()

array([ True, False])

In [85]:
# Replace 'False' with 'No' and 'True' with 'Yes' in the 'PaperlessBilling' column for data1 


data1['PaperlessBilling'].replace(True, 'Yes', inplace=True)
data1['PaperlessBilling'].replace(False, 'No', inplace=True)
data1['PaperlessBilling'].unique()

array(['Yes', 'No'], dtype=object)

In [86]:
# checking uique values in  data2['PaperlessBilling']
data2['PaperlessBilling'].unique()

array(['Yes', 'No'], dtype=object)

# PaymentMethod
No cleaning needed because the 'PaymentMethod' columns in the datasets had similar attributes. No misssing values to fill


In [87]:
data1['PaymentMethod'].unique()

array(['Electronic check', 'Mailed check', 'Bank transfer (automatic)',
       'Credit card (automatic)'], dtype=object)

In [88]:
data2['PaymentMethod'].unique()

array(['Credit card (automatic)', 'Electronic check', 'Mailed check',
       'Bank transfer (automatic)'], dtype=object)

# MonthlyCharges
Converting the float to 2 d.p needed in data1. No misssing values to fill


In [89]:
# checking d.p in data1 values
data1['MonthlyCharges'].head()

0    29.850000
1    56.950001
2    53.849998
3    42.299999
4    70.699997
Name: MonthlyCharges, dtype: float64

In [90]:
#convert to 2 d.p

data1['MonthlyCharges'] = data1['MonthlyCharges'].round(2)
data1['MonthlyCharges'].head()

0    29.85
1    56.95
2    53.85
3    42.30
4    70.70
Name: MonthlyCharges, dtype: float64

In [91]:
# checking data2 to see the d.p values 
data2['MonthlyCharges'].head()

0     49.50
1     55.00
2     43.95
3     74.35
4    111.15
Name: MonthlyCharges, dtype: float64

# TotalCharges
TotalCharges has 5 missing values in data1 and datatypes are different

In [92]:
# preview TotalCharges from data1 
data1['TotalCharges'].head()

0      29.850000
1    1889.500000
2     108.150002
3    1840.750000
4     151.649994
Name: TotalCharges, dtype: float64

In [93]:
#convert to 2 d.p
data1['TotalCharges'] = data1['TotalCharges'].round(2)
data1['TotalCharges'].head()

0      29.85
1    1889.50
2     108.15
3    1840.75
4     151.65
Name: TotalCharges, dtype: float64

In [94]:
# check for null values
data1['TotalCharges'].isnull().sum()

5

In [95]:
# Drop rows with any missing values in the TotalCharges column

data1.dropna(subset='TotalCharges', inplace=True)
data1['TotalCharges'].isnull().sum()

0

In [96]:
# preview TotalCharges in data2
data2['TotalCharges'].head()

0     312.7
1    1046.5
2    2960.1
3     834.2
4    6953.4
Name: TotalCharges, dtype: object

In [97]:
# Convert the TotalCharges column in data2 to float datatype
data2['TotalCharges'] = pd.to_numeric(data2['TotalCharges'], errors='coerce').astype(float)

# Round the values to 2 d.p 
data2['TotalCharges'] = data2['TotalCharges'].round(2)
data2['TotalCharges'].head()

0     312.7
1    1046.5
2    2960.1
3     834.2
4    6953.4
Name: TotalCharges, dtype: float64

# Churn 
Churn has 1 missing value in data1.
Missing value will be dropped

In [98]:
data1['Churn'].isnull().sum()

1

In [99]:
data1['Churn'].unique()

array([False, True, None], dtype=object)

In [100]:
# Drop row with the missing value in the Churn column in data1
data1.dropna(subset='Churn', inplace=True)
data1['Churn'].isnull().sum()

0

In [101]:
data1['Churn'].unique()

array([False, True], dtype=object)

In [102]:
# replace 'True' with 'Yes' and 'False' with 'No'

data1['Churn'].replace(True, 'Yes', inplace=True)
data1['Churn'].replace(False, 'No', inplace=True)
data1['Churn'].unique()

array(['No', 'Yes'], dtype=object)

In [103]:
# cross check Churn column in data2
data2['Churn'].unique()

array(['No', 'Yes'], dtype=object)